src/error.rs

```diff
@@ -38,28 +45,33 @@ impl Error {
     /// Encountered an unexpected situation which should not happen in practice.
     pub const UNEXPECTED: Error = Self::new_internal(2);
 
-    /// Codes below this point represent OS Errors (i.e. positive i32 values).
-    /// Codes at or above this point, but below [`Error::CUSTOM_START`] are
-    /// reserved for use by the `rand` and `getrandom` crates.
+    #[deprecated]
+    #[doc(hidden)]
     pub const INTERNAL_START: u32 = 1 << 31;
 
-    /// Codes at or above this point can be used by users to define their own
-    /// custom errors.
+    #[deprecated]
+    #[doc(hidden)]
     pub const CUSTOM_START: u32 = (1 << 31) + (1 << 30);
 
+    /// Internal errors can be in the range of 2^16..2^17
+    const INTERNAL_START2: RawOsError = 1 << 16;
+    /// Custom errors can be in the range of 2^17..(2^17 + 2^16)
+    const CUSTOM_START2: RawOsError = 1 << 17;
+
     /// Creates a new instance of an `Error` from a particular OS error code.
     ///
     /// This method is analogous to [`std::io::Error::from_raw_os_error()`][1],
     /// except that it works in `no_std` contexts and `code` will be
-    /// replaced with `Error::UNEXPECTED` if it isn't in the range
-    /// `1..Error::INTERNAL_START`. Thus, for the result `r`,
-    /// `r == Self::UNEXPECTED || r.raw_os_error().unsigned_abs() == code`.
+    /// replaced with `Error::UNEXPECTED` in unexpected cases.
     ///
     /// [1]: https://doc.rust-lang.org/std/io/struct.Error.html#method.from_raw_os_error
     #[allow(dead_code)]
-    pub(super) fn from_os_error(code: u32) -> Self {
-        match NonZeroU32::new(code) {
-            Some(code) if code.get() < Self::INTERNAL_START => Self(code),
+    pub(super) fn from_os_error(code: RawOsError) -> Self {
```

Comment:
If we want to handle all the error negating/checking in here, this could be:

On UEFI:
```rust
if code & UEFI_ERROR_FLAG  == 0 {
  Self::UNEXPECTED
} else {
  Self(NonZeroRawOsError::new(code).unwrap())
}
```

On platforms where `RawOsError` is expected to be positive:
```rust
match code.checked_neg() {
  None => Self::ERRNO_NOT_POSITIVE,
  Some(e) if e >= 0 => Self::ERRNO_NOT_POSITIVE,
  Some(e) => Self(NonZeroRawOsError::new(e).unwrap())
}
```
and on platforms where `RawOsError` is expected to be negative:
```rust
if code >= 0 {
  Self::UNEXPECTED
} else {
  Self(NonZeroRawOsError::new(code).unwrap())
}
```
then we just undo things in `raw_os_error`

Reply:
Hm. The problem with this approach is that we would have to manually list all targets which use positive (or negative) codes in the `error` module and do it twice.

But we kind of already do it with `#[cfg(not(target_os = "solid_asp3"))]`...

Reply:
I tweaked the code a bit. I left the `ERRNO_NOT_POSITIVE` check in the `last_os_error` function, but otherwise it's close to your proposal.